File: Source/CSharp/Expressive/Expressive/Functions/String/SubstringFunction.cs
using Expressive.Expressions;

namespace Expressive.Functions.String
{
    internal class SubstringFunction : FunctionBase
    {
        #region FunctionBase Members

        public override string Name
        {
            get
            {
                return "Substring";
            }
        }

        public override object Evaluate(IExpression[] parameters, ExpressiveOptions options)
        {
            this.ValidateParameterCount(parameters, 3, 3);

            object value = parameters[0].Evaluate(Variables);

            if (value == null)
            {
                return null;
            }

            // Safely handle the text input, if not text then call ToString.
            string text = null;
            if (value is string)
            {
                text = (string)value;
            }
            else
            {
                text = value.ToString();
            }

            int startIndex = (int)parameters[1].Evaluate(Variables);
            int length = (int)parameters[2].Evaluate(Variables);
            
            return text?.Substring(startIndex, length);
        }

        #endregion
    }
}

File: Source/Java/Expressive/src/main/java/com/bijington/expressive/functions/string/SubstringFunction.java
package com.bijington.expressive.functions.string;

import com.bijington.expressive.exceptions.ExpressiveException;
import com.bijington.expressive.expressions.IExpression;
import com.bijington.expressive.functions.FunctionBase;
import com.bijington.expressive.helpers.Convert;
import com.bijington.expressive.helpers.Strings;

/**
 * Created by shaun on 20/12/2016.
 */
public class SubstringFunction extends FunctionBase {
    @Override
    public String getName() {
        return "Substring";
    }

    @Override
    public Object evaluate(IExpression[] parameters) throws ExpressiveException {
        this.validateParameterCount(parameters, 3, 3);

        Object value = parameters[0].evaluate(this.getVariables());
        Object startIndexValue = parameters[1].evaluate(this.getVariables());
        Object length = parameters[2].evaluate(this.getVariables());

        if (value == null || startIndexValue == null || length == null) return null;

        String stringValue;
        if (value.getClass().equals(String.class)) {
            stringValue = String.class.cast(value);
        }
        else {
            stringValue = value.toString();
        }

        int startIndex = Convert.toInteger(startIndexValue);
        int totalLength = Convert.toInteger(length);

        return stringValue.substring(startIndex, startIndex + totalLength);
    }
}
